rust/headless-client/src/ipc_service.rs

```diff
@@ -192,42 +211,72 @@ impl Handler {
         })
     }
 
-    // Infallible so that we only give up on an IPC client explicitly
-    async fn run(&mut self) {
+    /// Run the event loop to communicate with an IPC client.
+    ///
+    /// If the IPC service needs to terminate, we catch that from `signals` and send
+    /// the client a hint to shut itself down gracefully.
+    ///
+    /// The return type is infallible so that we only give up on an IPC client explicitly
+    async fn run(&mut self, signals: &mut signals::Terminate) -> HandlerOutcome {
         loop {
-            let event = {
-                // This borrows `self` so we must drop it before handling the `Event`.
-                let cb = pin!(self.cb_rx.recv());
-                match future::select(self.ipc_rx.next(), cb).await {
-                    future::Either::Left((Some(Ok(x)), _)) => Event::Ipc(x),
-                    future::Either::Left((Some(Err(error)), _)) => {
-                        tracing::error!(?error, "Error while deserializing IPC message");
-                        continue;
+            let event = poll_fn(|cx| {
+                if let Poll::Ready(()) = signals.poll_recv(cx) {
+                    // `recv` on signals is cancel-safe.
+                    Poll::Ready(Event::Terminate)
+                } else if let Poll::Ready(result) = pin!(&mut self.ipc_rx).poll_next(cx) {
```

Comment:
For complex return types from a poll like this, I typically `match` right away so there is only one layer of pattern matching.

Reply:
Like this? (Comparing old and new)
<img width="552" alt="image" src="https://github.com/user-attachments/assets/b49f4cf7-22d4-4846-b1bc-baaf5e749017">
Maybe if I did `use Poll::Ready`. Otherwise the lines are a bit long.

Reply:
Yeah! The lines are long but the benefit is that the compiler makes an exhaustiveness check for you so you can glance over it in most cases.